File: kattis/inf237/graph_I/bowserpipes.py
from heapq import heappush

class Room:
    def __init__(self, index, next_room):
        self.is_entry = True
        self.index = index
        self.next_room = next_room
        
        if next_room == -1:
            self.end_room = index
            self.dist = 0
        else:
            self.end_room = -1
            self.dist = -1
    

    def __lt__(self, other):
        return (self.dist, self.index).__lt__((other.dist, other.index))


    def __eq__(self, other):
        return self.dist == other.dist and self.index == other.index


n = int(input())
rooms = list()
entry_points = [[] for _ in range(n)] 


def dfs(index, is_first_call):
    current = rooms[index]
    if not is_first_call:
        current.is_entry = False
    if current.dist != -1:
        return current

    path = dfs(current.next_room, False)
    current.end_room = path.end_room
    current.dist = path.dist + 1

    return current


def initialize():
    global rooms
    rooms = [Room(i, int(n)) for i, n in enumerate(input().strip().split(' '))]

    for i in range(n):
        dfs(i, True)

    for i in range(n):
        current = rooms[i]
        if current.is_entry:
            heappush(entry_points[current.end_room], current)


def mario_path(start):
    end = rooms[start].end_room 
    return entry_points[end][0].index


initialize()
a = int(input())
for entry in [int(p) for p in input().strip().split(' ')]:
    print(mario_path(entry))

File: kattis/inf237/graph_I/bowserpipes.cpp
#include <bits/stdc++.h>

using namespace std;

typedef struct Room {
  int dist;
  int index;
  int next;
  int end;
  bool isEntry; 
} Room;

vector<Room> rooms;

Room& dfs(int index, bool isFirstCall) {
  Room& current = rooms[index];
  if (!isFirstCall) {
    current.isEntry = false;
  }

  if (current.dist != -1) {
    return current;
  }

  Room& path = dfs(current.next, false);
  current.end = path.end;
  current.dist = path.dist + 1;

  return current;
}

bool cmp(const Room& a, const Room& b) {
  if (a.dist == b.dist) {
    return a.index < b.index;
  } else {
    return a.dist < b.dist;
  }
}

vector<vector<Room>> initialize(int n) {
  int val;
  for (int i = 0; i < n; i++) {
    cin >> val; 
    if (val == -1) {
      rooms.push_back({0, i, -1, i, true});
    } else {
      rooms.push_back({-1, i, val, -1, true});
    }
  }

  for (int i = 0; i < n; i++) {
    dfs(i, true);
  }

  vector<vector<Room>> entryPoints(n);
  for (int i = 0; i < n; i++) {
    Room current = rooms[i];
    if (current.isEntry) {
      entryPoints[current.end].push_back(current);
    }
  }

  return entryPoints;
}

int marioPath(int start, vector<vector<Room>>& entryPoints) {
  int end = rooms[start].end; 
  vector<Room>& entries = entryPoints[end];
  return min_element(entries.begin(), entries.end(), cmp)->index;
}

int main() {
  int n, a, entry;
  cin >> n;

  vector<vector<Room>> entryPoints = initialize(n);
  cin >> a;
  for (int i = 0; i < a; i++) {
    cin >> entry;
    cout << marioPath(entry, entryPoints) << '\n';
  }
}
